# Exercise 4.5 Data Consistency Checks

This script contains:

1. Import libraries
2. Data consistency checks on products dataset
3. Data consistency checks on orders dataset
4. Export dataframes

# 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Path to IC folders
path = r"C:\Users\alabaro\OneDrive - DAI\Documents\Aug_2023 Instacart Basket Analysis"

In [16]:
# importing orders cvs
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [17]:
# importing products cvs
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 02. Data Consistency Checks on df_prods

In [18]:
# checking for mixed types
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [19]:
# checking for null values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [21]:
# look for null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [22]:
# null values in product_name column
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [23]:
# checking shape of df_prods
df_prods.shape

(49693, 5)

In [24]:
# creating new dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
# checking shape of df_prods_clean - there should be 16 less rows
df_prods_clean.shape

(49677, 5)

In [26]:
# checking for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [27]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
# checking shape of df_prods_clean
df_prods_clean.shape

(49677, 5)

In [29]:
# removing duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
# checking new shape of df_prods_clean_no_dups - 5 are removed
df_prods_clean_no_dups.shape

(49672, 5)

# 03. Data Consistency on df_ords

In [31]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

For orders_days_of_the_week - does not have values more than 7 days of the week
For ourder_hour_of_the_day - does not have values exceding 24 hours
For days_since_prior_order - min and max values are 1 and 3, but the 50% value is 7. There might be something off in this column with both min and max values so low.  

In [32]:
# checking for mixed types - no mixed data types detected
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [33]:
# checking for null values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Days since prior order have 206,209 missing values. This is probably indicative of customers who have only visited the store once and do not have a prior order date. This is less than about 6% of the data. 

In [35]:
# checking shape of df_ords
df_ords.shape

(3421083, 7)

In [37]:
# finding median of days_since_prior_order - 7.0 days
df_ords.median()

Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

In [38]:
# replacing missing values with median of 7.0 days
df_ords['days_since_prior_order'].fillna(7.0, inplace=True)

In [39]:
# checking for null values after replacing, there are no null values now
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

To address the missing values in days_since_prior_order, I replaced them with the median value. Comparing the mean of 1 and the median of 7, this shows that there is a lot of skew in the data. For this reason, it's better to use the median to replace null values. 

In [40]:
# checking for duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [41]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates were found. 

In [44]:
df_ords.head(15)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
1            1   2398795        1             2                   3   
2            2    473747        1             3                   3   
3            3   2254736        1             4                   4   
4            4    431534        1             5                   4   
5            5   3367565        1             6                   2   
6            6    550135        1             7                   1   
7            7   3108588        1             8                   1   
8            8   2295261        1             9                   1   
9            9   2550362        1            10                   4   
10          10   1187899        1            11                   4   
11          11   2168274        2             1                   2   
12          12   1501582        2             2                   5   
13          13   1901567        2             3                   1   
14          14    738281        2             4                   2   

    order_hour_of_day  days_since_prior_order  
0                   8                     7.0  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     7.0  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0

# 04. Exporting Dataframes

In [45]:
# export df_ords to prepared data folder
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [47]:
# export df_ords to prepared data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))